assets/data-port/stepper/index.jsx

```diff
@@ -0,0 +1,102 @@
+import classnames from 'classnames';
+
+/**
+ * Helper method to update the steps when moving to the next one.
+ *
+ * @param  {Array} steps  The current steps.
+ * @return {Array}        The steps after moving.
+ */
+const moveToNext = ( steps ) => {
+	const newSteps = [ ...steps ];
+
+	for ( let i = 0; i < newSteps.length; i++ ) {
+		if ( newSteps[ i ].isActive ) {
+			newSteps[ i ].isComplete = true;
```

Comment:
@renatho 
A good strategy is always to return the new complete object. In this case, it would not be a problem now because our values are only primitive types, but if we create a new level in the object in some moment, it could be a problem.

When we are cloning the object in the `...steps`, it clones the object, creating a new reference, but it'll use the reference of the props.

Here we could use something like (I'm sorry by repeating the validations and some error, but only for example):

```javascript
newSteps[ i ] = {
  ...newSteps[ i ]
  isComplete: true,
  isActive: ! ( i + 1 < newSteps.length )
};
newSteps[ i + 1 ] = {
  ...newSteps[ i + 1 ],
  isActive: i + 1 < newSteps.length
};
```

Also, I think in this case, as we are modifying different items in each interaction (what prevents us to use the map), we could use the `Array.reduce`. But would not be a problem to keep as a `for`

-----

@gkaragia 
> When we are cloning the object in the ...steps, it clones the object, creating a new reference, but it'll use the reference of the props.

Why do we need to create new objects for the elements of the array though? I think that React is doing a 'shallow' reference comparison and since we create a new array every time it would be enough for React to understand that there was an update in the state and rerender.

> we could use the Array.reduce

Not sure if its appropriate to use reduce here since the semantics of the method is to apply an accumulator on the array to produce a single result. In our case we read the array to create an other array with elements of the same type so it could cause confusion.

Reply:
Oops! I answered in the original PR. Copying to here:

> Why do we need to create new objects for the elements of the array though? I think that React is doing a 'shallow' reference comparison and since we create a new array every time it would be enough for React to understand that there was an update in the state and rerender.

Yes! You are correct! It's a shallow comparison. So because that is important to "recreate" all the tree for the prop that we are changing.

I'm sorry that I didn't use a good example. Let's imagine that our component will be something like:

```javascript
<div>
  {steps.map( step =>
    <Step step={step}>
  )}
</div>
```

For the React, the `step` in this case, will never update because of the shallow comparison. Even we are updating the props, the reference for each object of the array is the same.

An example to run in the console:
```javascript
const steps = [
    { active: false }
];

const newSteps = [ ...steps ];

console.log( steps[0].active === newSteps[0].active ); // true

newSteps[0].active = true;

console.log( steps[0].active ); // true
```

Just reinforcing, probably it would not happen to the current code, but it would be easy to create an issue while refactoring.

Reply:
@renatho Sorry! Missed that 🤦‍♂️

Reply:
In the example that you describe, won't the `<div>` be part of a component which rerenders though?

What I am saying is even if we add your example in the `SenseiImportPage` element it would still get rerendered since React is going to call the render method of the parent element. You can try this by defining a new element:

```
const Text = ( { step } ) => {
	if (step.isActive) {
		return (
			<p>
				{step.key} Active
			</p>
		);
	} else {
		return (
			<p>
				{step.key} Inactive
			</p>
		);
	}
};
```
And then add the following line below the buttons (looping also works):
```
<Text key={steps[1].key} step={steps[1]}/>
```

In any case I think that even if there is an actual case where rerender wouldn't work, I think that we should avoid adding the code beforehand and handle it when the case arises.

Reply:
@jom No, You didn't miss that. I added that comment after you transport the comments. 😅

Reply:
@gkaragia, Ah, yes! If it's a children, it'll re-render while rendering the parent. But complementing this test, you can add to the `Test` component the following effect:

```javascript
useEffect( () => {
	console.log( JSON.stringify( step ) );
}, [ step ] );
```

You will see that this effect will run only in the first render.

Another example would be adding the `memo` in the `Text` component. Like: `const Text = memo( ( { step } ) => ... )`. In this case it won't update the component on the parent render because it will compare the new and previous props before to re-render.

In summary, what I mean is that usually when we mutate objects in React, we can fall into several traps. So, usually when we are changing an object, it is a good practice always to return new objects instead of mutating them.

Reply:
Hey @renatho

> You will see that this effect will run only in the first render.

That's because in the snippet the dependencies are not properly defined. Normally `useEffect` should have a dependency on the steps array (same is true for `useMemo`:

```
useEffect( () => {
	console.log( JSON.stringify( step[0] ) );
}, [ steps ] );
```

In general what you describe is a concept in Redux, as described [here](https://redux.js.org/recipes/structuring-reducers/immutable-update-patterns). To my understanding this has no application in normal React however, so I would prefer to avoid the extra complication if possible. In this case the impact is really small but things get ugly on using state of 2 or more levels deep. Redux itself uses immer to avoid this issue.

Reply:
Hey @gkaragia! 

> That's because in the snippet the dependencies are not properly defined. Normally useEffect should have a dependency on the steps array (same is true for useMemo):

Actually, the ideal hook dependencies are the dependencies that we're using in the scope. If we used the whole `steps`, for example, it would run for all steps even if they didn't change. I'm sorry to insist, but it really can cause some issues. In other cases too, like to send this value through a callback to a parent component, or using context, and so on. In some cases, issues caused by mutability are very hard to catch.

I think as our scope is not so big yet, we could solve it with spread to avoid a new dependency. But if you prefer, I agree to add a dependency for this. Another dependency for this could be https://github.com/kolodny/immutability-helper that I think is a little cleaner for simple cases, but both Immutability helper, and Immer, would be good for me.

Reply:
I can see what you mean but this still seems like something that should be considered when the case arises. In any case I think that we dwelled on this too much 😄 1a41f33efc60f90a04e94d02fd5e0cdf0af612c7